Allow the state of dialog Virtual Machine to be cloned, serialized and restored

naNuke opened the pull request:
_This is a solution for #181 and #60, not directly but in a reasoning also solution for #86_

Hello! I gave it a shot and created save / restore system for Virtual Machine State. I am using Googles Protobuf to handle serialization because it was already part of Yarns library so why not. The usage is very simple and straight forward _(I hope)_.

Summary of the changes I made:
---

- Move State out of VirualMachine and implement StateOption and StateValue, all on top of .proto
- Changed Stack inside State to List, but it still behaves like Stack through the public methods. _(I did this because its less pain to serialize List while keeping rest of the application intact.)_
- Implement other protobuf classes `StateOption` and `StateValue` to handle serialization
- Implement serializers `JsonSerializedState` and `BinarySerializedState` (protobuf) classes implementing `ISerializedState` interface. These are used to handle Json string or Protobuf bytestring and provide access to both original data through `GetData()` method and deserialized `State` class through `Deserialize()` method.
- Implement Getters and Setters across `VirtualMachine` and `Dialogue` classes.
- Implement `TryGetState()` and `TrySetState()` for checking if State can be get or set
- Add additional overloads for `DialogRunner.StartDialogue()` to allow starting from saved state.

Public method descriptions:
---
`dialogue.TryGetState()`
--> True or False + debug messages if current State is not prepared to be saved (program didnt run at all and Node wasnt set)

`dialogue.GetStateClone()`
--> Gets clone of the object. This method allows users to use their own serialization solution.

`dialogue.GetStateJsonSerialized()` `dialogue.GetStateBinarySerialized()`
--> Gets object used to access both serialized and deserialized data. It has `GetData()` method and `Deserialize()` method, `GetData()` gets either Json string or Protobuf ByteString while `Deserialize()` returns the resulting object by parsing the data

`dialogue.TrySetState()`
--> True or False + debug messages if there is something wrong with the State

`dialogue.SetState()`
--> This method will override current State, including current node etc. It accepts both `ISerializedState` interface and `State` as a source of State object.

`dialogueRunner.StartDialogue()`
--> This method has two additional overloads to allow starting dialogue directly from `ISerializedState` interface or `State`

---

So, I did a lot of testing and everything seems to work fine, but I am expecting requests/commentary on this. If this solution will be satisfactory I will commit some actual tests.

Pull Request open to discussion :)

Thank you!

_updated description based on latest commit 64f5862_

naNuke commented on Unity/Assets/YarnSpinner/Scripts/DialogueRunner.cs:
```diff
@@ -353,14 +353,6 @@ void Start ()
             
         }
 
-        private void Update()
```
This was forgotten testing thing that I forgot to remove before doing first commit.

naNuke commented on Unity/Assets/YarnSpinner/Examples/Space/Space.unity:
```diff
@@ -1894,6 +1899,31 @@ PrefabInstance:
       propertyPath: onDialogueEnd.m_PersistentCalls.m_Calls.Array.data[0].m_Arguments.m_BoolArgument
       value: 1
       objectReference: {fileID: 0}
+    - target: {fileID: 5935829475398715334, guid: 4ec3df5875e7347beb6176b545b8fb39,
```
When loading a VM State that is supposed to show options, the UIBehaviour would not hide continue button, so I have added disabling OnOptionsStart call

McJones commented on YarnSpinner/VirtualMachine.cs:
```diff
@@ -117,60 +116,209 @@ internal enum TokenType {
+        // Using this flag to return expected VM State programCounter.
+        // depends on when the State is requested, mostly it would be after
+        // an instruction has been processed which increments the counter.
+        private bool instructionProcessed = false;
```
I am slightly confused as to the purpose of this flag.
It appears to switch between on and off based on how far through running an instruction you are but so is it for solving the issue of when you want to save the state?

If that is the case I don't really think that specifically is worth supporting, you have control over when you save and load the state so handling edge cases like that just feels like a debugging nightmare waiting to happen.

Or have I misunderstood its purpose?
Could you please add some more information as to this flags purpose?

naNuke replied:
It is something I came up with while testing the solution, at the moment the only way to get correct programCounter is to call GetState before instruction is processed. Which is through UIBehaviour events like OnLineStart, OnOptionsStart etc. 

If I am just starting at the screen at a dialogue currently shown on the screen and press save button or use menu to save my state, the programCounter at that point will be +1 while stack and values will be same.

Loading such state breaks the VM.

I haven't thought of better solution but will try to improve it.

I tried moving programCounter++ in continue method before the RunInstruction() but that did break some code as well so it might need more work.

McJones commented on YarnSpinner/VirtualMachine.cs:
```diff
@@ -117,60 +116,209 @@ internal enum TokenType {
-            /// Pop a value from the stack
-            public Value PopValue() {
-                return stack.Pop ();
+    public partial class VMState
```
Any particular reason why you've renamed the class from State to VMState?

naNuke replied:
Just a naming convention I thought would be better to understand, since it is no longer nested in Virtual machine. Making custom State somewhere doesn't sound as self explanatory as VMState. Should I pull it back?

McJones commented on YarnSpinner/yarn_spinner.proto:
```diff
@@ -3,131 +3,147 @@ package Yarn;
+    int32 programCounter = 2; //as per protobuf documentation: For numeric types, the default value is zero.
+    repeated VMStateOption currentOptions = 3;
+    repeated VMStateValue stack = 4;
+}
```
I don't feel that this should go inside the same proto file.
The rest of the messages in this file are for what become assets, the VM state isn't an asset.

Would you mind pulling this out into its own definition file?

naNuke replied:
Ok thank you, I wasn't entirely sure if I should make new proto for those. Will move it.

McJones commented on YarnSpinner/VirtualMachine.cs:
```diff
@@ -117,60 +116,209 @@ internal enum TokenType {
+                return this._deserialized;
+            }
+
+            JsonParser parser = JsonParser.Default;
```
Is there any reason you've used the JsonParser over the normal protobuf serialiser/deserialiser?

naNuke replied:
On this I was thinking to provide SerializedVMState classes for JSON, XML, and proto, but I wasn't sure if it should be Yarn interest to provide multiple options so I decided to make JSON default as most people will be able to work with that.

If it would be preferred to default to proto that's fine, or should I make all 3 (2) options?

naNuke replied:
Thank you :), I will address the problems as soon as I can.

McJones commented:
This is an excellently documented pull request and looks close to being ready to merge in.

I do have some questions about some decisions you've made (left as in-line comments), if you could address those we can probably get this pull request accepted soon.

naNuke commented:
Thank you, I will address the issues as I can :)

naNuke commented:
Resolved all issues (I think?) left in the comments, removed the hacky solution for synchronizing currentInstruction with programCounter and it now uses separate property instead which feels better indeed.

Renamed State and all related classes back to State, sorry for overthinking that, its not like there will any different States around in near future.

Also implemented Protobuf serializer next to Json one.

... and little clean up

naNuke commented:
Most probably last commit from my end unless there are more requests or questions about the code.

I have further Improve serializers, added Try methods and added tests.

McJones commented:
This is all looking good, I am going to merge this into a new feature branch.
Because we are making some big changes to how lines are handled in 1.1 and only want to handle that, this won't get merged into that release but should be good to go with 1.2

Thanks so much for this.

desplesda commented:
Picking this thread up again: is this in a state that we'd be ok with merging into the 1.2 release?

naNuke commented:
It's behind a quite few commits so most probably needs manual fixing here and there, but as far as I been following, mainly the State() class might need to be looked at. And ofcourse documentation.

Let me know if I should update the PR to latest state of develop branch. :)
